File: apps/admin_backend/src/core/audit/v3/frame.ts
import { AUDIT_V3_ERROR_CODE as CODE, type AuditV3ErrorCode, type Result } from './types';

const MAX = 1_048_576;
const FIELD_STATE = { ABSENT: 'absent', NULL: 'null', TEXT: 'text' } as const;
export type FieldState =
  | { readonly state: typeof FIELD_STATE.ABSENT }
  | { readonly state: typeof FIELD_STATE.NULL }
  | { readonly state: typeof FIELD_STATE.TEXT; readonly value: string };
export interface AuditV3FrameFields {
  readonly user_id: string;
  readonly resolved_action: string;
  readonly device_id: string;
  readonly timestamp: string;
  readonly sequence_no: string;
  readonly prev_hash: string;
  readonly metodo_autorizacion: FieldState;
  readonly usuario_autorizador_id: FieldState;
}

class Failure extends Error {
  constructor(readonly code: AuditV3ErrorCode, readonly offset: number) { super(code) }
}
const fail = (code: AuditV3ErrorCode, offset: number): never => { throw new Failure(code, offset) };
const scalar = (value: string): boolean => !/[\ud800-\udfff]/.test(value);

export function buildAuditV3Frame(fields: AuditV3FrameFields, canonicalMetadata: Buffer): Result<Buffer> {
  try {
    const parts: Buffer[] = [Buffer.from('OFA3', 'ascii')];
    const required = [fields.user_id, fields.resolved_action, fields.device_id, fields.timestamp, fields.sequence_no, fields.prev_hash];
    const add = (tag: number, bytes: Buffer, index: number): void => {
      if (bytes.length > MAX) fail(CODE.FRAME_TOO_LARGE, index);
      const header = Buffer.alloc(5); header[0] = tag; header.writeUInt32BE(bytes.length, 1); parts.push(header, bytes);
    };
    required.forEach((value, index) => {
      if (typeof value !== 'string' || !scalar(value)) fail(CODE.FRAME_INVALID, index);
      if ((index === 1 || index === 3) && value.length === 0) fail(CODE.FRAME_INVALID, index);
      if (index === 4 && (!/^(?:0|[1-9][0-9]*)$/.test(value) || BigInt(value) > 4_294_967_295n)) fail(CODE.FRAME_INVALID, index);
      add(0x01, Buffer.from(value, 'utf8'), index);
    });
    [fields.metodo_autorizacion, fields.usuario_autorizador_id].forEach((field, offset) => {
      const index = offset + 6;
      if (!field || typeof field !== 'object') fail(CODE.FRAME_INVALID, index);
      if (field.state === FIELD_STATE.ABSENT) add(0x00, Buffer.alloc(0), index);
      else if (field.state === FIELD_STATE.NULL) add(0x03, Buffer.alloc(0), index);
      else if (field.state === FIELD_STATE.TEXT && typeof field.value === 'string' && scalar(field.value)) add(0x01, Buffer.from(field.value, 'utf8'), index);
      else fail(CODE.FRAME_INVALID, index);
    });
    if (!Buffer.isBuffer(canonicalMetadata)) fail(CODE.FRAME_INVALID, 8);
    add(0x02, canonicalMetadata, 8);
    const frame = Buffer.concat(parts);
    if (frame.length > MAX) fail(CODE.FRAME_TOO_LARGE, 9);
    return { ok: true, value: frame };
  } catch (error: unknown) {
    if (error instanceof Failure) return { ok: false, error: { code: error.code, offset: error.offset } };
    throw error;
  }
}

File: apps/pos_app/lib/core/audit/v3/frame.dart
import 'dart:convert';
import 'dart:typed_data';

import 'types.dart';

const _maxBytes = 1048576;

sealed class FieldState {
  const FieldState();
  const factory FieldState.absent() = _Absent;
  const factory FieldState.nil() = _Nil;
  const factory FieldState.text(String value) = _Text;
}

final class _Absent extends FieldState { const _Absent(); }
final class _Nil extends FieldState { const _Nil(); }
final class _Text extends FieldState {
  final String value;
  const _Text(this.value);
}

final class AuditV3FrameFields {
  final String userId;
  final String resolvedAction;
  final String deviceId;
  final String timestamp;
  final String sequenceNo;
  final String prevHash;
  final FieldState metodoAutorizacion;
  final FieldState usuarioAutorizadorId;

  const AuditV3FrameFields({
    required this.userId,
    required this.resolvedAction,
    required this.deviceId,
    required this.timestamp,
    required this.sequenceNo,
    required this.prevHash,
    required this.metodoAutorizacion,
    required this.usuarioAutorizadorId,
  });

  AuditV3FrameFields copyWith({
    String? userId, String? resolvedAction, String? deviceId, String? timestamp,
    String? sequenceNo, String? prevHash, FieldState? metodoAutorizacion, FieldState? usuarioAutorizadorId,
  }) => AuditV3FrameFields(
    userId: userId ?? this.userId, resolvedAction: resolvedAction ?? this.resolvedAction,
    deviceId: deviceId ?? this.deviceId, timestamp: timestamp ?? this.timestamp,
    sequenceNo: sequenceNo ?? this.sequenceNo, prevHash: prevHash ?? this.prevHash,
    metodoAutorizacion: metodoAutorizacion ?? this.metodoAutorizacion,
    usuarioAutorizadorId: usuarioAutorizadorId ?? this.usuarioAutorizadorId,
  );
}

final class _Part {
  final int tag;
  final Uint8List bytes;
  final int index;
  const _Part(this.tag, this.bytes, this.index);
}

bool _validUnicode(String value) {
  final units = value.codeUnits;
  for (var i = 0; i < units.length; i++) {
    final unit = units[i];
    if (unit >= 0xd800 && unit <= 0xdbff) {
      if (++i >= units.length || units[i] < 0xdc00 || units[i] > 0xdfff) return false;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      return false;
    }
  }
  return true;
}

AuditV3Result<Uint8List> buildAuditV3Frame(
  AuditV3FrameFields fields,
  Uint8List canonicalMetadata,
) {
  final values = [fields.userId, fields.resolvedAction, fields.deviceId, fields.timestamp, fields.sequenceNo, fields.prevHash];
  final parts = <_Part>[];
  for (var index = 0; index < values.length; index++) {
    final value = values[index];
    if (!_validUnicode(value) || ((index == 1 || index == 3) && value.isEmpty)) {
      return AuditV3Failure(AuditV3Error(auditV3FrameInvalid, index));
    }
    if (index == 4 && (!RegExp(r'^(?:0|[1-9][0-9]*)$').hasMatch(value) || BigInt.parse(value) > BigInt.from(0xffffffff))) {
      return const AuditV3Failure(AuditV3Error(auditV3FrameInvalid, 4));
    }
    parts.add(_Part(0x01, Uint8List.fromList(utf8.encode(value)), index));
  }
  for (final (index, field) in [(6, fields.metodoAutorizacion), (7, fields.usuarioAutorizadorId)]) {
    switch (field) {
      case _Absent(): parts.add(_Part(0x00, Uint8List(0), index));
      case _Nil(): parts.add(_Part(0x03, Uint8List(0), index));
      case _Text(:final value):
        if (!_validUnicode(value)) return AuditV3Failure(AuditV3Error(auditV3FrameInvalid, index));
        parts.add(_Part(0x01, Uint8List.fromList(utf8.encode(value)), index));
    }
  }
  parts.add(_Part(0x02, canonicalMetadata, 8));
  var size = 4;
  for (final part in parts) {
    if (part.bytes.length > _maxBytes) return AuditV3Failure(AuditV3Error(auditV3FrameTooLarge, part.index));
    size += 5 + part.bytes.length;
  }
  if (size > _maxBytes) return const AuditV3Failure(AuditV3Error(auditV3FrameTooLarge, 9));

  final output = Uint8List(size);
  output.setAll(0, const [0x4f, 0x46, 0x41, 0x33]);
  final data = ByteData.sublistView(output);
  var offset = 4;
  for (final part in parts) {
    output[offset] = part.tag;
    data.setUint32(offset + 1, part.bytes.length, Endian.big);
    output.setAll(offset + 5, part.bytes);
    offset += 5 + part.bytes.length;
  }
  return AuditV3Success(output);
}
